README.md
Alpha status

ChangeLog
CHANGES
=======

* Reintroduce support to download latest build from a specific tag
* Fix KojiDownloader to be usable as a source provider
* Fix Reposync to clean dnf metadata after download is done
* kojidownload command now can define download directory
* Tox: enable py36 env by default
* Pylint fixes
* Fix PEP8 warnings
* Fix repository file parsing to work on both 2.7 and 3.2+
* Fix tox to run unit tests
* Various fixes
* Initial state


sinkhole/tests/test_repo.py
"""Tests for sinkhole.repo module
"""
from sinkhole.repo import RepositoryFile

import dnf
import pytest


class TestRepositoryFile(object):
    """Tests for RepositoryFile class
    """
    def test_parse_repofile(self):
        """Test RepositoryFile .repo file parsing
        """
        data = """
[hguemar-rdo-packager]
name=Copr repo for rdo-packager owned by hguemar
baseurl=https://copr-be.cloud.fedoraproject.org/results/hguemar/rdo-packager/fedora-27-x86_64/
type=rpm-md
skip_if_unavailable=True
gpgcheck=1
gpgkey=https://copr-be.cloud.fedoraproject.org/results/hguemar/rdo-packager/pubkey.gpg
repo_gpgcheck=0
enabled=1
enabled_metadata=1"""
        base = dnf.Base()
        conf = base.conf
        repofile = RepositoryFile(conf, data)
        repos = repofile.get_repos()
        assert len(repos) == 1

    def test_raise_error_non_existing_config(self):
        """Test if RepositoryFile fails when provided
        non existing .repo file
        """
        repofn = "toto.conf"
        base = dnf.Base()
        conf = base.conf
        with pytest.raises(IOError,
                           message="File {:s} does not exist".format(repofn)):
            RepositoryFile(conf, repofn)


sinkhole/tests/test_util.py
"""Tests for sinkhole.util
"""
from sinkhole.util import filter_pkgs


class FakeRPM(object):
    """Test utility class to fake RPM packages
    objects in dnf API
    """
    def __init__(self, name, source_name):
        self.name = name
        self.source_name = source_name

    def __hash__(self):
        return hash((self.name, self.source_name))

    def __eq__(self, other):
        return self.name == other.name and \
            self.source_name == other.source_name


class TestFilterPkgs(object):
    """Tests for filter_pkgs function
    """
    def test_filter_pkgs_1(self):
        """ Test filtering
        """
        pkgs = {FakeRPM("python2-requests-2.18.4-1.fc27.noarch.rpm",
                        "python-requests-2.18.4-1.fc27.src.rpm")}
        patterns = ["python-requests.*"]
        res = filter_pkgs(pkgs, patterns)
        assert res == pkgs

    def test_filter_pkgs_2(self):
        """ Test filtering
        """
        pkgs_old = {FakeRPM("python2-requests-2.18.4-1.fc27.noarch.rpm",
                            "python-requests-2.18.4-1.fc27"
                            ".src.rpm")}
        pkgs = pkgs_old | {FakeRPM("grin-1.2.1-15.fc27.noarch.rpm",
                                   "grin-1.2.1-15.fc27.src.rpm")}
        patterns = ["python-requests.*"]
        res = filter_pkgs(pkgs, patterns)
        assert res == pkgs_old


sinkhole/tests/test_config.py
"""Tests for sinkhole.config
"""
from sinkhole.config import Config

from six import StringIO


class TestConfig(object):
    """Tests for Config class
    """
    def test_shared_state(self):
        """Check that state is shared across instances of Config
        """
        config1 = Config()
        config1.workers = 4
        config1.output_dir = "/tmp/some_dir"
        config2 = Config()
        assert config1.workers == config2.workers == 4
        assert config1.output_dir == config2.output_dir == "/tmp/some_dir"

    def test_read_config(self):
        """Check that configuration file are properly parsed
        """
        data = StringIO("""[main]
workers = 8
output_dir = output/
releasever = 28
basearch = x86_64

[sources.fedora]
type=repofile
repofile=sources/fedora.repo
constraints=sources/fedora_constraints
""")
        config = Config.read_config(data)
        assert config.workers == 8
        assert config.output_dir == "output/"
        assert config.default_substitutions["releasever"] == "28"
        assert config.default_substitutions["basearch"] == "x86_64"


sinkhole/util.py
""" Bunch of utilities
"""

import os.path
import re
import shutil

import requests


def filter_pkgs(pkgs, patterns):
    """ Filter a list of packages according patterns

    Args:
        pkgs (list): List of packages to filter
        patterns (list): List of patterns

    Returns:
        set: a filtered set of package
    """
    result = set()
    for pattern in patterns:
        regex = re.compile(r"^{}$".format(pattern))
        result |= {pkg for pkg in pkgs if regex.match(pkg.source_name)}

    return result


def download_packages(url, destdir):
    """ direct packages download utility

    Args:
        url: url of the package to download
        destdir: destination directory

    returns:
        bool: success or failure
    """
    pkg_name = url.split("/")[-1]
    path = os.path.join(destdir, pkg_name)
    if os.path.exists(path):
        print("{} already exists. SKIPPED".format(path))
        return True
    response = requests.get(url, allow_redirects=True, stream=True)
    with open(path, 'wb') as fdst:
        shutil.copyfileobj(response.raw, fdst)
    return True


sinkhole/config.py
""" Config file management
"""

import six
from six.moves import configparser

if six.PY2:
    configparser.ConfigParser.read_file = \
        lambda self, x: configparser.ConfigParser.readfp(self, x)


class Config(object):
    """Configuration object

    Uses Borg pattern to keep shared state across instances
    """
    _shared_state = {
        "_workers": 4,
        "_output_dir": "tmp/",
        "_sources": [],
        "_default_substitutions": {},
        "_masher_opts": "",
        "_masher_workers": 1
    }

    def __init__(self):
        self.__dict__ = self._shared_state

    @property
    def workers(self):
        """Property: workers (Number of processes)
        """
        return self._workers

    @workers.setter
    def workers(self, workers):
        """Property(setter): workers (Number of processes)
        """
        self._workers = workers

    @property
    def output_dir(self):
        """Property: output_dir (Output directory)
        """
        return self._output_dir

    @output_dir.setter
    def output_dir(self, output_dir):
        """Property(setter): output_dir (Output directory)
        """
        self._output_dir = output_dir

    @property
    def sources(self):
        """Property: sources (sources providers)
        """
        return self._sources

    @sources.setter
    def sources(self, sources):
        """Property(setter): sources (sources providers)
        """
        self._sources = sources

    @property
    def default_substitutions(self):
        """Property: default_substitutions (Yum repo file substitutions)
        """
        return self._default_substitutions

    @property
    def masher_opts(self):
        """Property: masher_opts (createrepo options)
        """
        return self._masher_opts

    @masher_opts.setter
    def masher_opts(self, opts):
        """Property(setter): masher_opts (createrepo options)
        """
        self._masher_opts = opts

    @property
    def masher_workers(self):
        """Property: masher_workers (createrepo workers)
        """
        return self._masher_workers

    @masher_workers.setter
    def masher_workers(self, nb):
        """Property(setter): masher_workers (createrepo workers)
        """
        self._masher_workers = int(nb)

    @classmethod
    def read_config(cls, fname):
        """Read config file

        Args:
          fname: either filename or file-like object pointing to configuration

        Returns:
            Config: Config object
        """
        config = cls()
        conf = configparser.ConfigParser()
        fobj = fname if hasattr(fname, "readline") else open(fname, "r")
        conf.read_file(fobj)

        workers = conf.getint("main", "workers")
        output_dir = conf.get("main", "output_dir")
        config.workers = workers
        config.output_dir = output_dir
        config.default_substitutions["releasever"] = \
            conf.get("main", "releasever")
        config.default_substitutions["basearch"] = conf.get("main", "basearch")

        try:
            config.masher_opts = conf.get("masher", "opts")
            config.masher_workers = conf.getint("masher", "workers")
        except configparser.NoSectionError:
            print("Warning: Masher has not been configured")
        except configparser.NoOptionError:
            print("Warning: Masher options has not been defined."
                  "Using default parameters")

        sources = [section for section in conf.sections()
                   if section.startswith("sources")]
        config.sources = []
        for source in sources:
            provider = {k: conf.get(source, k) for k in conf.options(source)}
            provider["name"] = source
            config.sources.append(provider)
        return config


sinkhole/kojidl.py
""" Koji interaction
"""

import os.path

import koji
import six
import yaml

from sinkhole.config import Config
from sinkhole.util import download_packages


def setup_kojiclient(profile):
    """Setup koji client session
    """
    opts = koji.read_config(profile)
    for key, value in six.iteritems(opts):
        opts[key] = os.path.expanduser(value) if isinstance(value, str) \
                    else value
    kojiclient = koji.ClientSession(opts['server'], opts=opts)
    # FIXME: who cares about authentication???
    # Moreover kerberos authentication can be an annoyance
    if opts['authtype'] == "ssl" or \
       (os.path.isfile(opts['cert']) and opts['authtype'] is None):
        kojiclient.ssl_login(opts['cert'], None, opts['serverca'])
    else:
        print('SKIP authentication')
    return kojiclient


class KojiDownloader(object):
    """ Download builds from koji
    """
    def __init__(self, profile, builds=None, arches=None):
        self.profile = profile
        self.koji = setup_kojiclient(self.profile)
        self._arches = arches
        self._builds = self.builds(builds)

    def builds(self, builds_):
        """ Defines builds to be downloaded
        """
        res = []
        for build in builds_:
            nvr = self._retrieve_nvr(build)
            if nvr:
                res.append(nvr)
            else:
                print("Error: {} not Found in {}".format(build). self.profile)
        return res

    def _retrieve_nvr(self, build):
        """ Retrieve NVR from Koji instance
        """
        if "@" not in build:
            return build

        build, tag = build.split("@")
        info = None
        if tag == "latest":
            pkg_id = self.koji.getPackageID(build)
            info = self.koji.listBuilds(pkg_id)
        else:
            info = self.koji.listTagged(tag, package=build)

        nvr = info[0]['nvr'] if info else None
        return nvr

    @classmethod
    def build(cls, info):
        """Build a KojiDownloader instance
        """
        kojid = None
        profile = info["profile"]
        constraints_file = info["constraints"]
        with open(constraints_file, "r") as cfile:
            constraints = yaml.load(cfile, Loader=yaml.Loader)
            builds = constraints["builds"]

        kojid = cls(profile, builds)
        return kojid

    def run(self):
        """ Execute downloads
        """
        urls = []
        pathinfo = koji.PathInfo(topdir=self.koji.opts['topurl'])
        config = Config()
        for build in self._builds:
            try:
                info = self.koji.getBuild(build)
                rpms = self.koji.listRPMs(buildID=info['id'],
                                          arches=self._arches)
                fnames = [pathinfo.rpm(rpm) for rpm in rpms]
                urls += [pathinfo.build(info) + '/' + fname
                         for fname in fnames]
            except Exception:
                print('SKIPPED: build {} does not exists'.format(build))
        for url in urls:
            download_packages(url, config.output_dir)


sinkhole/source.py
""" Source provider factory
"""
from collections import defaultdict

from sinkhole.config import Config
from sinkhole.kojidl import KojiDownloader
from sinkhole.repo import Reposync


class Noop(object):
    """ Default provider for unrecognized source
    """
    @classmethod
    def build(cls, info):
        """ Build source provider
        """
        print("Unrecognized source {} (type: {})".format(info["name"],
                                                         info["type"]))
        return None


SOURCES = {"repofile": Reposync, "koji": KojiDownloader}
SOURCES = defaultdict(lambda: Noop, SOURCES)


class SourceBuilder(object):
    """ Build a source from config
    """
    @staticmethod
    def build():
        """ Build all sources
        """
        config = Config()
        sources = []
        for info in config.sources:
            typ = info["type"]
            source = SOURCES[typ].build(info)
            if source:
                sources.append(source)
        return sources


sinkhole/command.py
#! /usr/bin/env python

""" Module containing command-line interface
"""

import argparse

from sinkhole.config import Config
from sinkhole.kojidl import KojiDownloader
from sinkhole.masher import RepoMasher
from sinkhole.repo import Reposync
from sinkhole.source import SourceBuilder


def main():
    """ Main entrypoint for sinkhole
    """
    # top-level parser
    parser = argparse.ArgumentParser(description="sinkhole")
    parser.add_argument("--config", action="store")
    subparsers = parser.add_subparsers(dest="subcommand")

    # reposync
    parser_reposync = subparsers.add_parser("reposync",
                                            help="Sync repositories")
    parser_reposync.add_argument("--destdir", action="store", required=True)
    parser_reposync.add_argument("--repofile", action="append", dest="repofns",
                                 required=True)
    parser_reposync.add_argument("--include", action="append", default=None)
    parser_reposync.add_argument("--exclude", action="append", default=None)

    # kojidownload
    parser_koji = subparsers.add_parser("kojidownload",
                                        help="Download builds from Koji")
    parser_koji.add_argument("--profile", action="store", default="koji",
                             help="Koji instance to retrieve builds from")
    parser_koji.add_argument("--arch", action="store", nargs="+",
                             default=None)
    parser_koji.add_argument("--builds", action="store", nargs="+",
                             required=True)
    parser_koji.add_argument("--destdir", action="store", required=True)

    # Run appropriate command
    args = parser.parse_args()

    config = Config.read_config(args.config) if args.config else Config()

    if args.subcommand == "reposync":
        config.output_dir = args.destdir
        reposync = Reposync(args.repofns, args.include, args.exclude)
        reposync.run()
    elif args.subcommand == "kojidownload":
        config.output_dir = args.destdir
        kojid = KojiDownloader(args.profile, args.builds, args.arch)
        kojid.run()
    else:
        sources = SourceBuilder.build()
        for source in sources:
            source.run()
        rm = RepoMasher(config.output_dir)
        rm.run()


if __name__ == '__main__':
    main()


sinkhole/repo.py
""" RPM repositories management
"""

from functools import partial
from multiprocessing import Pool
import os
import shutil

import dnf
from dnf.conf.parser import substitute
import six
from six.moves import configparser
import yaml

from sinkhole.config import Config
from sinkhole.util import (download_packages, filter_pkgs)

if six.PY2:
    # ConfigParser.read_string has been added in Python 3.2
    import io
    configparser.ConfigParser.read_string = \
        lambda self, x: configparser.ConfigParser.readfp(self, io.BytesIO(x))


class RepositoryFile(object):
    """ RepositoryFile parses a .repo file
    """
    def __init__(self, conf, repofn):
        self.repofile = repofn
        self.repos = []
        self._parse_repofile(repofn, conf)

    def _parse_repofile(self, repofn, conf):
        config = configparser.ConfigParser()
        if not os.path.isfile(repofn):
            # failover on a string
            try:
                config.read_string(repofn)
            except Exception:
                raise IOError("File {:s} does not exists".format(repofn))
        else:
            config.read(repofn)
        for section in config.sections():
            repo = dnf.repo.Repo(section, conf)
            for key, value in config.items(section):
                if key in ["name", "baseurl", "enabled",
                           "gpgkey", "gpgcheck"]:
                    setattr(repo, key, value)
                elif key == "metalink":
                    setattr(repo, key, substitute(value, conf.substitutions))
            self.repos.append(repo)

    def get_repos(self):
        """ Returns a list of repositories from .repo file

        Returns:
            list: list of repositories
        """
        return self.repos


class Reposync(object):
    """ Reposync allows to sync repositories with packages filtering
    Repositories to sync are specified by feeding this class with your classic
    .repo files.
    """
    def __init__(self, repofns=None,
                 include_pkgs=None,
                 exclude_pkgs=None):
        self.include_pkgs = include_pkgs
        self.exclude_pkgs = exclude_pkgs
        self.repofns = repofns if repofns is not None else []
        self.base = dnf.Base()
        self.conf = self.base.conf
        config = Config()
        self.conf.substitutions['releasever'] = \
            config.default_substitutions["releasever"]
        self.conf.substitutions['basearch'] = \
            config.default_substitutions["basearch"]
        self.conf.cachedir = os.path.join(Config().output_dir, "cache")
        self.repos = self.base.repos

    @property
    def substitutions(self):
        """ Return yum/dnf substitutions dict
        """
        return self.conf.substitutions

    @classmethod
    def build(cls, info):
        """Build an instance of Reposync
        """
        reposync, include_pkgs, exclude_pkgs = None, None, None
        repofile = info["repofile"]
        if "constraints" in info:
            constraints_file = info["constraints"]
            with open(constraints_file, "r") as cfile:
                constraints = yaml.load(cfile, Loader=yaml.Loader)
                include_pkgs = constraints["includes"]
                exclude_pkgs = constraints["excludes"]

        reposync = cls([repofile],
                       include_pkgs=include_pkgs,
                       exclude_pkgs=exclude_pkgs)

        for sub in ["releasever", "basearch"]:
            if sub in info:
                reposync.substitutions[sub] = info[sub]
        return reposync

    def run(self):
        """ Do the repo syncing
        """
        self._setup_repos()
        self.base.fill_sack()
        available_pkgs = self.base.sack.query().available().run()
        download_pkgs = self._filter_download_pkgs(available_pkgs)
        #
        # sinkhole --repofile fedora.repo --destdir tmp2
        # 23.27s user 2.14s system 1% cpu 28:19.24 total
        download_pkgs = [pkg.remote_location() for pkg in download_pkgs]

        config = Config()
        with Pool(config.workers) as pool:
            pool.map(partial(download_packages,
                             destdir=config.output_dir),
                     download_pkgs)
        # 3161
        # sinkhole --repofile fedora.repo --destdir tmp2
        # 699.75s user 195.08s system 14% cpu 1:41:17.52 total
        # self.base.download_packages(download_pkgs,
        #                             MultiFileProgressMeter(fo=sys.stdout))
        self._cleanup_dnf_artefacts()

    def _cleanup_dnf_artefacts(self):
        """ Delete dnf metadata artefacts
        """
        shutil.rmtree(self.conf.cachedir)

    def _setup_repos(self):
        """Parse repository files
        """
        for repofn in self.repofns:
            repofile = RepositoryFile(self.conf, repofn)
            repositories = repofile.get_repos()
            for repository in repositories:
                self.repos.add(repository)

    def _filter_download_pkgs(self, pkgs):
        """ Private method that returns a list of packages to download

        Args:
            pkgs (list): packages to filter

        Returns:
            list: packages to download
        """
        included = set(pkgs) if not self.include_pkgs else \
            filter_pkgs(pkgs, self.include_pkgs)
        excluded = set() if not self.exclude_pkgs else \
            filter_pkgs(pkgs, self.exclude_pkgs)
        pkgs = list(included - excluded)
        return pkgs
